### ISSUE TITLE
safe_eval is not safe

### ISSUE THREAD

**Issue by cedk** [odoo/odoo#17391]
Here is a report of a security issue that was reported to the Odoo security team but they considered it as not one.
Here is a copy of the initial report:

> Hi,
> 
> The safe_eval is not safe against long running computation.
> For example, this simple python statement takes all the CPU resource and
> grows to use all the available memory:
> 
>     9**9**9**9**9
> 
> As safe_eval is used in many places to evaluate external statement, this
> can be used by an attacker to exhaust easily the resources of the Odoo
> server.
> 
> Regards,

The main response of Odoo security team was that it is mitigated by the system limits (--limit* configuration).
But the system limits protect only other services running on the same server, against Odoo bad behavior (consuming all the CPU or memory) but it does not protect Odoo itself.
For me, the issue is in the design by pretending that safe_eval is safe.
Indeed a good solution will be to limit the sets of expressions that can be evaluated and ensure their evaluations do not cost too much resources. But Odoo wants to be able to evaluate any expression even the most expensive one.

In conclusion, safe_eval should not be used to evaluate expression coming from untrusted sources. But it is used in many places to evaluate expression from user.

**Comment by odony** [comment]
For the record, here is a summary of our position:
- We do recognize that in the modern information security world, attacks and abuse are increasing concerns. Users are exposed to more and more threats on the Internet every day, and software vendors must be prepared to do their part.
- However, the purpose of Odoo's `safe_eval` has always been to implement a sandbox that blocks access to unauthorized data and unauthorized Python code. Its job is to keep your data safe, not to protect your system resources from Denial of Service attacks. You can always recover from a DoS attack. A data breach is irreversible, with long-term consequences, so this is our foremost concern.
- In addition, from a technical standpoint it does not make sense to try to implement such a "resource protection" at the `safe_eval` level, because:
  * Determining whether an arbitrary program is going to exhaust resources or run forever is impossible. In computer theory, this is known as the [Halting Problem](https://en.wikipedia.org/wiki/Halting_problem).
  * Even if it was possible to implement it, such a protection would take wrong decisions and block legit cases which simply happen to be expensive. Users don't want that.
- That said, we do address the DoS risk. The protection of system resources is instead implemented with the --limit* configuration parameters [1]. Contrary to the `safe_eval` mechanism, this is done at the process level, the right place for taking system-preserving decisions. And because this is highly dependent on the deployment environment (hardware) and the business requirements of the users, these parameters require proper configuration by the server administrator. We have a basic security checklist for Odoo deployment [here](https://www.odoo.com/documentation/10.0/setup/deploy.html#security), by the way.
- Finally, one could argue that we should entirely drop the `safe_eval` feature if it turns out that it could be *abused* to cause a denial of service, when malicious users are given elevated privileges. We consider this view rather short-sighted, because on this account we should drop most other features! Unsurprisingly, actions accomplished by users in the system affect the world (resources) *in some way*. Users can send emails, they can populate database tables, they can print report, upload files, and so on. Any of these features could be similarly abused by repeatedly using them until *some* resource is exhausted (DISK, RAM, CPU, IP reputation, ...). Resources are always finite. 
Should we drop all those features? No, there would be no software left. What we do is we introduce tools to monitor and limit abuse instead. Of course it's not magical, and chances are that you will need to perform some manual mitigation if you're ever targeted by a DoS attack.
But removing `safe_eval` *would not* magically make you immune to that. What it *would do*, however, is cripple an important feature and reduce the flexibility of the system a lot.

Note: this discussion is mostly a concern for SaaS-like platform where untrusted users can obtain elevated privileges. In such environments, it is obviously critical for the SaaS vendor to implement appropriate isolation and resource limits in order to protect the platform and the users from other abusive users. Again, this is true with or without `safe_eval`.


[1] --limit-memory-soft, --limit-memory-hard, --limit-time-cpu and
--limit-time-real
